# Combining Data

Practicing combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.


# Exercise 1


In [1]:
import pandas as pd
df_rural = pd.read_csv('../data/rural_population_percent.csv',dtype=str,skiprows=3)
df_electricity = pd.read_csv('../data/electricity_access_percent.csv',dtype=str,skiprows=3)


df_rural=df_rural.drop(['Unnamed: 62'],axis=1)
df_electricity=df_electricity.drop(['Unnamed: 62'],axis=1)

combined_df=pd.concat([df_rural,df_electricity],ignore_index=True)


In [2]:
combined_df.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  ...  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254   49.27  49.285    49.3  ...   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195   90.89  90.574   90.25  ...   
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...   
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842   68.77  ...   
4  SP.RUR.TOTL.ZS   41.55  39.017  36.538  34.128  31.795  29.555  ...   

     2008    2009    2010    2011    2012    2013    2014    2015    2016  \
0  56.217  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696   
1  76.054  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868   
2  61.491  60.701  59.903    59.1  58.301   57.51  56.726   55.95  55.181   
3  50.009  48.924  47.837  46.753   45.67  44.617  43.591  42.593  41.624   
4  11.133  11.648  12.183   12.74  13.292  13.835  14.367  14.885  15.388   

     2017  
0  58.893  
1   72.43  
2  54.422  
3  40.684  
4  15.873  

[5 rows x 62 columns]

# Exercise 2 (Challenge)


Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc


In [3]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural = pd.read_csv('../data/rural_population_percent.csv',dtype=str,skiprows=3)
df_electricity = pd.read_csv('../data/electricity_access_percent.csv',dtype=str,skiprows=3)

df_rural=df_rural.drop(['Unnamed: 62'],axis=1)
df_electricity=df_electricity.drop(['Unnamed: 62'],axis=1)

df_rural=pd.melt(df_rural,id_vars=['Country Name', 'Country Code','Indicator Name','Indicator Code'], \
                 var_name='Year',value_name='Rural Value')

# TODO: drop any columns from the data frames that aren't needed
df_electricity=pd.melt(df_rural,id_vars=['Country Name', 'Country Code','Indicator Name','Indicator Code'], \
                 var_name='Year',value_name='Electricity Value')

df_rural=df_rural.drop(['Indicator Name','Indicator Code'],axis=1)
df_electricity=df_electricity.drop(['Indicator Name','Indicator Code'],axis=1)
# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year

# TODO: sort the results by country and then by year

df_combined = df_rural.merge(df_electricity,how='outer',on=['Country Name','Country Code','Year'])

In [4]:
df_combined.sort_values(by=['Country Name','Year'])

Country Name Country Code  Year Rural Value Electricity Value
1      Afghanistan          AFG  1960      91.779               NaN
265    Afghanistan          AFG  1961      91.492               NaN
529    Afghanistan          AFG  1962      91.195               NaN
793    Afghanistan          AFG  1963       90.89               NaN
1057   Afghanistan          AFG  1964      90.574               NaN
...            ...          ...   ...         ...               ...
30619     Zimbabwe          ZWE  Year         NaN              2013
30620     Zimbabwe          ZWE  Year         NaN              2014
30621     Zimbabwe          ZWE  Year         NaN              2015
30622     Zimbabwe          ZWE  Year         NaN              2016
30623     Zimbabwe          ZWE  Year         NaN              2017

[45936 rows x 5 columns]